### ISSUE TITLE
Applying fix leads to incorrect code

### ISSUE THREAD

**Issue by sztomi** [rust-lang/rust-clippy#11962]
### Summary

Applying fix leads to incorrect code

### Lint Name

clippy::ptr_arg

### Reproducer

I tried this code:

```rust
fn f(l: &Vec<char>) -> Vec<char> {
    vec![l[1], l[0]]
}

fn main() {
    let map = vec![vec!['.', '.']];

    let _: Vec<Vec<char>> = map
        .iter()
        .map(f)
        .collect();
}
```

I saw this happen:

```
warning: writing `&Vec` instead of `&[_]` involves a new object where a slice will do
 --> src/main.rs:1:9
  |
1 | fn f(l: &Vec<char>) -> Vec<char> {
  |         ^^^^^^^^^^ help: change this to: `&[char]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
  = note: `#[warn(clippy::ptr_arg)]` on by default
```

I expected to see this happen: Either a better suggestion (not sure what) or no suggestion at all. 

Applying this fix leads to a compiler error:

```rust
fn f(l: &[char]) -> Vec<char> {
    vec![l[1], l[0]]
}

fn main() {
    let map = vec![vec!['.', '.']];

    let _: Vec<Vec<char>> = map
        .iter()
        .map(f)
        .collect();
}
```

```
❯ cargo b     
   Compiling repro v0.1.0 (/home/tamas/tmp/repro)
error[E0631]: type mismatch in function arguments
   --> src/main.rs:10:14
    |
1   | fn f(l: &[char]) -> Vec<char> {
    | ----------------------------- found signature defined here
...
10  |         .map(f)
    |          --- ^ expected due to this
    |          |
    |          required by a bound introduced by this call
    |
    = note: expected function signature `fn(&Vec<char>) -> _`
               found function signature `for<'a> fn(&'a [char]) -> _`
note: required by a bound in `map`
   --> /home/tamas/.rustup/toolchains/1.72.1-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:803:12
    |
800 |     fn map<B, F>(self, f: F) -> Map<Self, F>
    |        --- required by a bound in this associated function
...
803 |         F: FnMut(Self::Item) -> B,
    |            ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Iterator::map`

error[E0599]: the method `collect` exists for struct `Map<Iter<'_, Vec<char>>, fn(&[char]) -> Vec<char> {f}>`, but its trait bounds were not satisfied
  --> src/main.rs:11:10
   |
8  |       let _: Vec<Vec<char>> = map
   |  _____________________________-
9  | |         .iter()
10 | |         .map(f)
11 | |         .collect();
   | |         -^^^^^^^ method cannot be called due to unsatisfied trait bounds
   | |_________|
   | 
   |
  ::: /home/tamas/.rustup/toolchains/1.72.1-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:61:1
   |
61 |   pub struct Map<I, F> {
   |   -------------------- doesn't satisfy `_: Iterator`
   |
   = note: the following trait bounds were not satisfied:
           `<for<'a> fn(&'a [char]) -> Vec<char> {f} as FnOnce<(&Vec<char>,)>>::Output = _`
           which is required by `Map<std::slice::Iter<'_, Vec<char>>, for<'a> fn(&'a [char]) -> Vec<char> {f}>: Iterator`
           `for<'a> fn(&'a [char]) -> Vec<char> {f}: FnMut<(&Vec<char>,)>`
           which is required by `Map<std::slice::Iter<'_, Vec<char>>, for<'a> fn(&'a [char]) -> Vec<char> {f}>: Iterator`
           `Map<std::slice::Iter<'_, Vec<char>>, for<'a> fn(&'a [char]) -> Vec<char> {f}>: Iterator`
           which is required by `&mut Map<std::slice::Iter<'_, Vec<char>>, for<'a> fn(&'a [char]) -> Vec<char> {f}>: Iterator`

Some errors have detailed explanations: E0599, E0631.
For more information about an error, try `rustc --explain E0599`.
error: could not compile `repro` (bin "repro") due to 2 previous errors
```

### Version

```text
rustc 1.76.0-nightly (06e02d5b2 2023-12-09)
binary: rustc
commit-hash: 06e02d5b259c1e88cbf0c74366d9e0a4c7cfd6d9
commit-date: 2023-12-09
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.5
```

### Additional Labels

* `I-suggestion-causes-error`

**Comment by PartiallyTyped** [comment]
It seems that Clippy doesn't take into consideration that the applied suggestion might fail. The lint is only using local information.

**Comment by PartiallyTyped** [comment]
I was expressing that the lint will only check the function, and will emit the error regardless of how it is used, it is then up to the programmer to adjust their code if they will accept the suggestion.
